### PR TITLE
Translations for core states

### PR DESCRIPTION
**Pull request by emlove**
This PR migrates the strings for the core states into the translation resources.

Notable points for discussion:
1. States which are displayed on a badge label currently have their own special processing which returns a shortened string to fit in the badge label. I've implemented the equivalent here with a separate state namespace just for states that can be shown as a badge label. I played with trying to achieve this with CSS instead, but I think the final solution for something that looks good requires a person to make the judgment call, especially considering that it will be different for each language. https://github.com/home-assistant/home-assistant-polymer/blob/aa389bf206cab1a0c638c250e43e0fbd27fb1ea1/src/components/entity/ha-state-label-badge.html#L139-L153
2. I moved the media player primaryText/secondaryText from the entity object to a mixin, so that it can access the localization methods. I think this is reasonable as well since this is really a display concern, not a entity model concern.
3. I moved the `computeStateState` util logic into StateCardDisplay, since that's the only place it's used now. We can keep it as a separate mixin if we want, but I'd rather just split it later if necessary.

### CODE REVIEW

**Comment by andrey-git** [comment]
Please don't merge until I got some time to review this :)

File: src/state-summary/state-card-display.html
Comment (balloob) [inline src/state-summary/state-card-display.html]: we should have the keys be under binary_sensor domain

File: src/translations/en.json
Comment (balloob) [inline src/translations/en.json]: Let's also put these under a domain. `state_badge.device_tracker.home` for example

File: src/util/media-player-model.html
Comment (balloob) [inline src/util/media-player-model.html]: This whole file should be converted to a class in the future.

File: src/util/hass-mixins.html
Comment (balloob) [inline src/util/hass-mixins.html]: Since this is the only place where it is used, we could have this logic live in the component and let the media player class just have a mediaTitle prop or something. If that is undefined, use the localize inside the component.

Comment (c727) [reply src/util/hass-mixins.html]: An the other hand you could use it in custom state cards or panels

File: src/translations/en.json
Comment (c727) [inline src/translations/en.json]: fix capitalzation

File: src/translations/en.json
Comment (andrey-git) [inline src/translations/en.json]: Does Lokalise support descriptions? Messages like this must be translated in context.

Comment (armills) [reply src/translations/en.json]: Agreed. I didn't see any way to add a description in our JSON source file, but we could at the least add a comment to the key when we upload it. It does seem that they support extra data for other formats, so maybe this is something that could be added: https://docs.lokalise.co/article/SOrbByx8ZT-custom-attributes

File: src/state-summary/state-card-display.html
Comment (andrey-git) [reply src/state-summary/state-card-display.html]: Maybe also under device_class?

File: src/translations/en.json
Comment (armills) [reply src/translations/en.json]: I think the binary sensor entries would be safe to separate out, since they only exist for display on the frontend.

However, I'm more concerned about the other domain states. We definitely have some states that are shared between domains, (i.e. STATE_OFF, STATE_ON), but more importantly, some of the domain states have been co-opted in a few odd places by other platforms. Some examples:

https://github.com/home-assistant/home-assistant/blob/2a28f7c4b506968255fea5a96bfe2c8f2ffd271d/homeassistant/components/media_player/roku.py#L138
https://github.com/home-assistant/home-assistant/blob/2a28f7c4b506968255fea5a96bfe2c8f2ffd271d/homeassistant/components/sensor/modem_callerid.py#L118
https://github.com/home-assistant/home-assistant/blob/2a28f7c4b506968255fea5a96bfe2c8f2ffd271d/homeassistant/components/sensor/mqtt_room.py#L74

I think that if we want to break these states out here, we should first do a backend refactoring to move these state constants into the domain components. Thoughts?

Comment (andrey-git) [reply src/translations/en.json]: I think it is OK for backend to share states.
The frontend should still have separate keys. Lokalise supports linking keys, so there is no need to translate several times unless there is real need.

File: src/state-summary/state-card-display.html
Comment (armills) [reply src/state-summary/state-card-display.html]: Hmm, I like the idea of nesting under device_class as well. We could get rid of this branching entirely, and just look it up by device class.

The only concern I'd have is the case here where gas, motion, occupancy, smoke, sound and vibration all have the same results. Would we add duplicates in the schema? (Not opposed, since maybe alternates make sense in other translations.)

File: src/translations/en.json
Comment (armills) [reply src/translations/en.json]: OK, I think I'm on board with that. So in our schema we'll have some duplicates across domains, but different languages might need different translations for different domain constants. 👍

File: src/components/entity/ha-state-label-badge.html
Comment (andrey-git) [inline src/components/entity/ha-state-label-badge.html]: Could be done in a separate PR, but we need to translate the units too.

Comment (armills) [reply src/components/entity/ha-state-label-badge.html]: Yep, we'll get there at some point 😄

File: src/util/hass-mixins.html
Comment (balloob) [reply src/util/hass-mixins.html]: Adam, what do you think of this? That way we can have 1 less mixin and keep all media player logic in the media player class.

Comment (armills) [reply src/util/hass-mixins.html]: Yeah, definitely agree, just have had a busy few days. I should be able to get to all of the comments tonight.

**Comment by armills** [comment]
I believe I've addressed all of the comments, but I've stumbled onto some things that need to be discussed.

1) Z-Wave is now introducing our first translation with a parameter substitution.
2) I've removed the behavior of falling back to the message in our localize function if the translation isn't found. It complicates the chaining of possible translations, and I think I was over-focused on the first use case when I added it. I'm thinking we should limit those fallbacks to where we explicitly want them.
3) I hit a snag where translations weren't being updated after the resources were loaded. I did some digging under the hood of the app-localize-behavior, and found that the behavior `localize` function is actually a computed function (which is how the recalculation is triggered after resources are updated). In order to not break the polymer recalculation, I've updated our implementation to be a computed function as well. Unfortunately that meant I had to rename our child class method to not conflict with the behavior. Open to opinions on a better name, or if we consider dropping it altogether and just build the keys where we use them, since we're no longer using the extra behavior of falling back to a message.

File: src/translations/en.json
Comment (c727) [inline src/translations/en.json]: [%key:state::switch::off%]

Comment (andrey-git) [reply src/translations/en.json]: We need to decide which file is the "fallback" one.
If this one is the fallback, then we can't have keys here.
Or we can use keys here and use translations/en.json as fallback.

Comment (armills) [reply src/translations/en.json]: Good points. I see three options we have here.

I don't like using translations/en.json myself, for a few reasons. First of all, it means it's going to complicate local development, since translations/en.json is only fetched from Lokalise. We could build against src/translations/en.json in dev mode, but it means we'd be getting the raw key references. It also means that after merging new translation keys in src, we'd be commiting the repo into a state with invalid key lookups, until the next round-trip to Lokalise.

I'm personally leaning towards src/translations/en.json. I'm not too worried about including duplicates in the base translation myself. I think that feature is going to be much more valuable for translators, especially since the app is natively written in English, we don't have to worry as much about wanting to change how "Off" looks everywhere. The nested placeholder example of cancel/proceed seems more useful, but I think in most of those cases we'd probably have variable placeholders anyway.

**Comment by andrey-git** [comment]
I think the use of keys is critical. It makes translations consistent by default and saves translation work.

I think we should use keys in src/ and use translations/en.js in prod/dev mode.

We can make a script for development purposes to replace keys with strings

**Comment by armills** [comment]
Sounds good, we can make that work.

As far as the actual schema goes, I'm thinking we should move the actual master on/off states into the [default schema](https://github.com/home-assistant/home-assistant-polymer/pull/575/files#diff-571ce32a17d6aa22fdd0942224a8ff44R13), to reflect the fact that a [toggle entity is a core hass concept](https://github.com/home-assistant/home-assistant/blob/62c1b542edf2ff24baf1749c57293af94c8413e5/homeassistant/helpers/entity.py#L357). The "on"/"off" for other domains can point to the default with a key reference.

For the others, (i.e. binary_sensor clear/detected, etc.) or any other duplicates that arise, I think we can just pick one to be the master.

File: src/util/hass-mixins.html
Comment (andrey-git) [inline src/util/hass-mixins.html]: I wonder if mixin is the right approach here,
This could be a function in utils - it doesn't need to be part of custom element prototype.
Then computeStateState could move back to utils too.

In any case, why introduce computeHaLocalize instead of overriding localize ?

File: src/state-summary/state-card-display.html
Comment (andrey-git) [inline src/state-summary/state-card-display.html]: For proper i18n we should have separate messages for with and without query_stage. (One can have the other as `[%key] ` by default.

File: gulp/tasks/translations.js
Comment (armills) [reply gulp/tasks/translations.js]: The only problem I have with that is it doesn't set the exit status of the process, so it won't fail the build. Google has been letting me down on this front.

File: src/util/hass-mixins.html
Comment (armills) [reply src/util/hass-mixins.html]: It needs to be at the component layer, because it's taking advantage of the Polymer app-localize-behavior, which only works because of the resources/language props which are part of the element state.

As far as why I couldn't just override localize, I'll copy my comment from above. Basically it boils down to the fact that Polymer's implementation works by rebinding the function when the resources become available, so in order to make use of that we need to do it the same way. If we also name our function `localize` here, we can't bind a computed property to the parent implementation.

> I hit a snag where translations weren't being updated after the resources were loaded. [I did some digging under the hood of the app-localize-behavior](https://github.com/PolymerElements/app-localize-behavior/blob/d3ff7ab958cc52a23b09c60faed55eced36dbc24/app-localize-behavior.html#L196), and found that the behavior `localize` function is actually a computed function (which is how the recalculation is triggered after resources are updated). In order to not break the polymer recalculation, I've updated our implementation to be a computed function as well. Unfortunately that meant I had to rename our child class method to not conflict with the behavior. Open to opinions on a better name, or if we consider dropping it altogether and just build the keys where we use them, since we're no longer using the extra behavior of falling back to a message.

File: src/state-summary/state-card-display.html
Comment (armills) [reply src/state-summary/state-card-display.html]: What do you think for the schema, something like this? I personally don't think we should be including keys we're not going to use, unless I'm not understanding what you mean.

```json
"zwave": {
  "default": {
      "initializing": "Initializing",
      "dead": "Dead",
      "sleeping": "Sleeping",
      "ready": "Ready"
  },
  "query_stage": {
      "initializing": "[%key:state::zwave::initializing%] ({query_stage})",
      "dead": "[%key:state::zwave::dead%] ({query_stage})",
      "sleeping": "[%key:state::zwave::sleeping%] ({query_stage})",
      "ready": "[%key:state::zwave::ready%] ({query_stage})"
  }
}
```

File: gulp/tasks/translations.js
Comment (armills) [reply gulp/tasks/translations.js]: Also have it on my list to come back here soon and clean up this pattern to match the rest of our gulp tasks.

File: src/state-summary/state-card-display.html
Comment (andrey-git) [reply src/state-summary/state-card-display.html]: Yes, this is what I meant. Except I don't think we need query_stage for ready and sleeping.

So maybe:
```json
"zwave": {
  "default": {
      "initializing": "Initializing",
      "dead": "Dead",
      "sleeping": "Sleeping",
      "ready": "Ready"
  },
  "query_stage": {
      "initializing": "[%key:state::zwave::default::initializing%] ({query_stage})",
      "dead": "[%key:state::zwave::default::dead%] ({query_stage})",
  }
}
```

Comment (armills) [reply src/state-summary/state-card-display.html]: 👍 Sounds good

File: src/state-summary/state-card-display.html
Comment (andrey-git) [inline src/state-summary/state-card-display.html]: Lets change ${domain} to zwave here and two lines below - it would be more readable

**Review by andrey-git — APPROVED** [review]

**Comment by armills** [comment]
👍 Great feedback. Should be good to go at this point.

The next test is to get some better testing infrastructure in place here. Too many easy places to slip up.

**Pull request merged**
